pkg/graveler/committed/apply.go

```diff
@@ -14,205 +14,296 @@ type ApplyOptions struct {
 	AllowEmpty bool
 }
 
-func addIntoDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType, n int) {
-	if d.Count != nil {
-		d.Count[typ] += n
-	}
-}
+// ReferenceType represents the type of the reference
 
-func incrementDiffSummary(d *graveler.DiffSummary, typ graveler.DiffType) {
-	addIntoDiffSummary(d, typ, 1)
+type applier struct {
```

Comment:
I don't understand why this is an object.  The only state it holds is a diff summary, which is anyway implemented as a function, and the two `have*` members.  It is created once, then its `apply` method is called -- with no parameters.  AFAICT it is just a function that has been represented as an object.  Having it as an object makes me have to worry about whether it does actually have important state.  Calls to functions that modify `have*` could be modeled as re-assigning some function-local state variables.  Alternatively, if there is a useful abstraction for "iterator-and-the-last-value-from-next", then let's implement that and put the `X` `Iterator` and `haveX` into a single class named something like `IteratorWithEnd`.

If the goal is to have a nice struct with all the used parameters in it then just have a struct, no need to attach methods to it.  But I would really question why passing a single parameter named `a` to a function is clearer than passing the named parameters to that function.  E.g. as it stands, the type signature for `addIntoDiffSumary` doesn't indicate that it actually changes only `a.summary`, but does not advance any iterators or write ranges.  Sure, the call `a.addIntoDiffSummary(typ, n)` is shorter than `addIntoDiffSummary(summary, type, n)`, but I claim that is just because it gives less information, and forces me to go read the code.

Also nit: `applier` sounds like Go naming for an interface.